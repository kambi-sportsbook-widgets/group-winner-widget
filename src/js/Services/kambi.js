import { offeringModule } from 'kambi-widget-core-library'

class KambiService {
  /**
   * Checks if given filter exists in current highlights.
   * @param {string} filter Filter to check
   * @returns {Promise.<boolean>}
   */
  static existsInHighlights(filter) {
    return offeringModule.getHighlight().then(response => {
      return !!response.groups.find(group =>
        group.pathTermId.match(
          new RegExp(`^/${filter.replace(/\/all/g, '')}(/all)*$`)
        )
      )
    })
  }

  /**
   * Fetches groups for given tournament.
   * @param {string} filter Tournament's filter
   * @param {number} criterionId Tournament's criterion identifier
   * @returns {Promise.<object[]>}
   */
  static getGroups(filter, criterionId) {
    return offeringModule
      .getEventsByFilter(`${filter}/all/competitions/`)
      .then(competitions => {
        return (
          competitions.events

            // must have exactly one betoffer
            .filter(event => event.betOffers && event.betOffers.length === 1)

            // criterion must match
            .filter(event => event.betOffers[0].criterion.id == criterionId)

            // set groupName
            .map(event => {
              const matches = event.event.englishName.match(/Group ([A-Z])/)

              if (!matches) {
                throw new Error(
                  `Cannot extract group letter from: ${event.event.englishName}`
                )
              }

              event.groupName = matches[1]

              return event
            })

            // sort based on groupName field
            .sort((a, b) => a.groupName.localeCompare(b.groupName))

            // remove the prelive event if there is one live of same groupName
            .filter((group, i, groups) => {
              if (group.betOffers[0].live) {
                return true
              }

              return !groups.find(
                g => g.groupName == group.groupName && g.betOffers[0].live
              )
            })

            // sort outcomes by odds value
            .map(group => {
              group.betOffers[0].outcomes.sort((a, b) => a.odds - b.odds)
              return group
            })
        )
      })
  }

  /**
   * Returns home team name of tournament's closest match.
   * @param {string} filter Tournament's filter
   * @returns {Promise.<string|null>}
   */
  static getNextMatchHomeName(filter) {
    return offeringModule
      .getEventsByFilter(`${filter}/all/matches/`)
      .then(matches => {
        const currentTime = Date.now()

        return matches.events
          .filter(
            m =>
              m.event.type === 'ET_MATCH' &&
              m.event.start &&
              m.event.start > currentTime
          )
          .sort((a, b) => a.event.start - b.event.start)
      })
      .then(matches => (matches.length > 0 ? matches[0].event.homeName : null))
  }
}

export default KambiService
