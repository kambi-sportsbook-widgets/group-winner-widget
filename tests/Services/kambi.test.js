import kambi from '../../src/js/Services/kambi';
import { coreLibrary, offeringModule, widgetModule } from 'kambi-widget-core-library';

const originalDateNow = window.Date.now;

jest.mock('kambi-widget-core-library', () => ({
   coreLibrary: {
      config: {
         routeRoot: ''
      }
   },
   offeringModule: {
      getHighlight: jest.fn(),
      getEventsByFilter: jest.fn()
   },
   widgetModule: {
      createFilterUrl: jest.fn()
   }
}));

const mockHighlightGroup = {pathTermId: '/test'};

describe('kambi.existsInHighlights()', () => {

   afterEach(() => {
      offeringModule.getHighlight.mockClear();
   });

   it('returns true if given filter is present in highlights', () => {
      offeringModule.getHighlight = jest.fn(() => new Promise(resolve =>
         resolve({ groups: [mockHighlightGroup] })));

      return kambi.existsInHighlights('test')
         .then((result) => {
            expect(result).toBe(true);
            expect(offeringModule.getHighlight).toHaveBeenCalledTimes(1);
         });
   });

   it('returns false if given filter is not present in highlights', () => {
      offeringModule.getHighlight = jest.fn(() => new Promise(resolve =>
         resolve({ groups: [mockHighlightGroup] })));

      return kambi.existsInHighlights('test_non_existing')
         .then((result) => {
            expect(result).toBe(false);
            expect(offeringModule.getHighlight).toHaveBeenCalledTimes(1);
         });
   });

});

describe('kambi.getGroups()', () => {

   afterEach(() => {
      offeringModule.getEventsByFilter = jest.fn();
   });

   it('finds next match team name correctly', () => {
      offeringModule.getEventsByFilter = jest.fn(() => new Promise(resolve =>
         resolve({
            events: [
               { betOffers: [] },
               { betOffers: [ {}, {} ] },
               { betOffers: [ { criterion: { id: 200 } } ]},
               { betOffers: [ { criterion: { id: 100 }, outcomes: [{odds: 20}, {odds: 10}] } ], event: { englishName: 'Group A (World Cup Qualification UEFA 2018)' }},
               { betOffers: [ { criterion: { id: 100 }, outcomes: [{odds: 20}, {odds: 10}] } ], event: { englishName: '(World Cup Qualification UEFA 2018) Group B' }},
               { betOffers: [ { criterion: { id: 100 }, live: true, outcomes: [{odds: 20}, {odds: 10}] } ], event: { englishName: '(World Cup Qualification UEFA 2018) Group C (World Cup Qualification UEFA 2018)' }}
            ]
         })));

      return kambi.getGroups('test', 100)
         .then((result) => {
            expect(result).toMatchSnapshot();
            expect(offeringModule.getEventsByFilter).toHaveBeenCalledTimes(1);
            expect(offeringModule.getEventsByFilter).toHaveBeenCalledWith('test/all/competitions/');
         });
   });

   it('throws exception when group name cannot be found', () => {
      offeringModule.getEventsByFilter = jest.fn(() => new Promise(resolve =>
         resolve({
            events: [
               { betOffers: [ { criterion: { id: 100 }, outcomes: [{odds: 20}, {odds: 10}] } ], event: { englishName: 'group A (World Cup Qualification UEFA 2018)' }}
            ]
         })));

      expect.assertions(1);

      return kambi.getGroups('test', 100)
         .catch(e => expect(e).toMatchSnapshot());
   });

});

describe('kambi.getNextMatchHomeName()', () => {

   beforeEach(() => {
      window.Date.now = () => 10;
   });

   afterEach(() => {
      offeringModule.getEventsByFilter.mockClear();
      window.Date.now = originalDateNow;
   });

   it('finds next match team name correctly', () => {
      offeringModule.getEventsByFilter = jest.fn(() => new Promise(resolve =>
         resolve({
            events: [
               { event: { type: 'ET_NOT_MATCH' } },
               { event: { type: 'ET_MATCH' } },
               { event: { type: 'ET_MATCH', start: 5 } },
               { event: { type: 'ET_MATCH', start: 15, homeName: 'Poland' } },
               { event: { type: 'ET_MATCH', start: 20, homeName: 'Sweden' } }
            ]
         })));

      return kambi.getNextMatchHomeName('test')
         .then((result) => {
            expect(result).toBe('Poland');
            expect(offeringModule.getEventsByFilter).toHaveBeenCalledTimes(1);
            expect(offeringModule.getEventsByFilter).toHaveBeenCalledWith('test/all/matches/');
         });
   });

   it('returns null when there is no upcoming matches', () => {
      offeringModule.getEventsByFilter = jest.fn(() => new Promise(resolve =>
         resolve({ events: [ ] })));

      return kambi.getNextMatchHomeName('test')
         .then(result => expect(result).toBe(null));
   });

});
