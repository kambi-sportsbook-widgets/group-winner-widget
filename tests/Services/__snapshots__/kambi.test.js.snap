// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`kambi.getGroups() finds next match team name correctly 1`] = `
Array [
  Object {
    "betOffers": Array [
      Object {
        "criterion": Object {
          "id": 100,
        },
        "outcomes": Array [
          Object {
            "odds": 10,
          },
          Object {
            "odds": 20,
          },
        ],
      },
    ],
    "event": Object {
      "englishName": "Group A (World Cup Qualification UEFA 2018)",
    },
    "groupName": "A",
  },
  Object {
    "betOffers": Array [
      Object {
        "criterion": Object {
          "id": 100,
        },
        "outcomes": Array [
          Object {
            "odds": 10,
          },
          Object {
            "odds": 20,
          },
        ],
      },
    ],
    "event": Object {
      "englishName": "(World Cup Qualification UEFA 2018) Group B",
    },
    "groupName": "B",
  },
  Object {
    "betOffers": Array [
      Object {
        "criterion": Object {
          "id": 100,
        },
        "live": true,
        "outcomes": Array [
          Object {
            "odds": 10,
          },
          Object {
            "odds": 20,
          },
        ],
      },
    ],
    "event": Object {
      "englishName": "(World Cup Qualification UEFA 2018) Group C (World Cup Qualification UEFA 2018)",
    },
    "groupName": "C",
  },
]
`;

exports[`kambi.getGroups() throws exception when group name cannot be found 1`] = `[Error: Cannot extract group letter from: group A (World Cup Qualification UEFA 2018)]`;
